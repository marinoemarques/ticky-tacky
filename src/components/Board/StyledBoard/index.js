import styled from 'styled-components'

const StyledBoard = styled.div`
    display: grid;
    grid-gap: 1px;
    background-color: Gray;
    grid-template-areas:
        "quare-0 square-1 square-2"
        "square-3 square-4 square-5"
        "square-6 square-7 square-8";
    grid-template-rows: 20vh 20vh 20vh;
    grid-template-columns: 20vh 20vh 20vh;
    margin: 2vh auto;
    height: calc(60vh + 2px);
    width: calc(60vc + 2px);
`
StyledBoard.displayName = 'StyledBoard'

export default StyledBoard
