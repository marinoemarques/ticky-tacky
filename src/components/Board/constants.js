export const INITIAL_STATE = {
  moves: []
}

export const SQUARE_PLAYED = 'SQUARE_PLAYED'
