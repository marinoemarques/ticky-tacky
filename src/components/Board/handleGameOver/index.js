import { length } from 'ramda'
import { isNilOrEmpty, isNotEmpty, isOdd } from 'ramda-adjunct'

import {
  GAME_OVER,
  MINIMUM_MOVES_TO_WIN,
  PLAYER_ONE,
  PLAYER_TWO
} from '../../../constants'
import getWinningPattern from '../../../utilities/getWinningPattern'

export default function handleGameOver (state, dispatch) {
  const { moves, outcome } = state
  const plays = length(moves)

  if (plays >= MINIMUM_MOVES_TO_WIN) {
    const patterns = getWinningPattern(moves)

    if (isNotEmpty(patterns) && isNilOrEmpty(outcome)) {
      const player = isOdd(plays) ? PLAYER_ONE : PLAYER_TWO

      dispatch({
        type: GAME_OVER,
        outcome: {
          player,
          patterns
        }
      })
    }
  }
}
