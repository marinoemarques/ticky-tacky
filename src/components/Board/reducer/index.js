import { SQUARE_PLAYED } from '../../constants'

export default function reducer (state, action) {
  switch (action.type) {
    case SQUARE_PLAYED:
      return {
        ...state,
        moves: [
          ...state.moves,
          action.square
        ]
      }
    default:
      return state
  }
}
