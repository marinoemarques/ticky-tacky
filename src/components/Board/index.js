import * as React from 'react'

import { times } from 'ramda'

import {
  INITIAL_STATE,
  NUMBER_OF_SQUARES,
  SQUARE_PLAYED
} from '../../constants'
import getPlayer from '../../utilities/getPlayer'
import Square from '../Square'

import handleGameOver from './handleGameOver'
import reducer from './reducer'
import StyledBoard from './StyledBoard'

const { useEffect, useReducer } = React

function makeSquares (state, dispatch) {
  const { outcome } = state

  return times(
    idx => {
      const player = getPlayer(state, idx)
      const onClick = player || outcome
        ? undefined
        : () => dispatch({ type: SQUARE_PLAYED, square: idx })

      return (
        <Square
          key={`square-${idx}`}
          area={`square-${idx}`}
          onClick={onClick}
          player={player}
        />
      )
    },
    NUMBER_OF_SQUARES
  )
}

export default function Board () {
  const [state, dispatch] = useReducer(reducer, INITIAL_STATE)

  useEffect(() => {
    handleGameOver(state, dispatch)
  })

  return (
    <StyledBoard>
      {makeSquares(state, dispatch)}
    </StyledBoard>
  )
}
