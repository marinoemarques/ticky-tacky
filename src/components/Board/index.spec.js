import * as React from 'react'

import Board from './'
import { shallow } from 'enzyme'

describe('Board', () => {
  it('matches the snapshot', () => {
    expect(toJson(shallow(<Board />))).toMatchSnapshot()
  })
})
