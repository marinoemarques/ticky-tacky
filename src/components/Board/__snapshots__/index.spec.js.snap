// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board matches the snapshot 1`] = `
<StyledBoard>
  <Square
    area="square-0"
  />
  <Square
    area="square-1"
  />
  <Square
    area="square-2"
  />
  <Square
    area="square-3"
  />
  <Square
    area="square-4"
  />
  <Square
    area="square-5"
  />
  <Square
    area="square-6"
  />
  <Square
    area="square-7"
  />
  <Square
    area="square-8"
  />
</StyledBoard>
`;
